packages/react-devtools-shared/src/devtools/views/Profiler/PerfInsight.js

```diff
@@ -0,0 +1,142 @@
+import {ProfilerContext} from '../Profiler/ProfilerContext';
+import {StoreContext} from '../context';
+
+import styles from './PerfInsight.css';
+
+type Props = {|
+  fiberID: number,
+|};
+
+export default function PerfInsight({fiberID}: Props) {
+  const {profilerStore} = useContext(StoreContext);
+  const {rootID, selectedCommitIndex} = useContext(ProfilerContext);
+
+  if (!enableProfilerPerfInsights) {
+    return null;
+  }
+
+  // TRICKY
+  // Handle edge case where no commit is selected because of a min-duration filter update.
+  // If the commit index is null, suspending for data below would throw an error.
+  // TODO (ProfilerContext) This check should not be necessary.
+  if (selectedCommitIndex === null) {
+    return null;
+  }
+
+  const {perfInsights} = profilerStore.getCommitData(
+    ((rootID: any): number),
+    selectedCommitIndex,
+  );
+
+  if (perfInsights === null) {
+    return null;
+  }
+
+  const perfInsight = perfInsights.get(fiberID);
+  if (perfInsight == null) {
+    return null;
+  }
+
+  const {
+    isFirstMount,
+    didContextChange,
+    didContextDeepChange,
+    didHooksChange,
+    didPropsChange,
+    didPropsDeepChange,
+    didStateChange,
+    didStateDeepChange,
+    nonMemoizedProps,
+  } = perfInsight;
+
+  if (isFirstMount) {
+    return (
+      <div className={styles.Component}>
+        <label className={styles.Label}>Why did this render?</label>
+        <div className={styles.Item}>
+          This is the first time the component rendered.
+        </div>
+      </div>
+    );
+  }
+
+  const insights = [];
+
+  if (!didContextChange && didContextDeepChange) {
+    insights.push(
+      <div key="context" className={styles.Item}>
+        • Context deep value changed
+      </div>,
+    );
+  }
+
+  // If context and hooks didn't change and  props/state are shallow equal,
+  // then this component needs basic memo.
+  if (didContextChange && !didHooksChange && didPropsChange && didStateChange) {
+    insights.push(
+      <div className={styles.Item}>
+        • Use React.PureComponent or React.Memo to avoid this component
+        rendering.
+      </div>,
+    );
+  }
+
+  if (!didPropsChange && didPropsDeepChange) {
+    insights.push(
+      <div className={styles.Item}>
+        • Use shouldComponentUpdate or React.Memo with deep equality to avoid
+        this component rendering.
+      </div>,
+    );
+  }
+
+  if (!didStateChange && didStateDeepChange) {
+    insights.push(
+      <div className={styles.Item}>
+        • Use shouldComponentUpdate with deep equality to avoid this component
+        rendering.
+      </div>,
+    );
+  }
+
+  if (nonMemoizedProps != null && nonMemoizedProps.length > 0) {
+    insights.push(
+      <div className={styles.Item}>
+        • Cant compare deep compare props:
+        {nonMemoizedProps.map(key => (
+          <span key={key} className={styles.Key}>
+            {key}
+          </span>
+        ))}
+        . These should be memoized.
```

Comment:
<img width="423" alt="Screen Shot 2021-11-09 at 3 06 09 PM" src="https://user-images.githubusercontent.com/29597/140996798-db5595e8-3485-41c6-b2c2-c5c4a6487f75.png">

I'm concerned about the use of the word "_should_" in this context.

Sometimes memoization "costs" more than not memoizing. (For example, if an object/function changes nearly every render anyway, then memoizing it strictly adds _more overhead_.)

In the case of the `children` prop, which is kind of special, this concern also seems relevant. Might we risk confusing people by telling them they "_should_" be memoizing children? In many cases this would not be an option.

Maybe we should weaken the wording here? e.g. "consider memoizing"